### PR TITLE
Moves sockets into the advanced worker

### LINKED ISSUE #7894: Performance: Move cluster based socket work into web worker (step 1)
**Is your feature request related to a problem? Please describe.**
- Updates to resources are received to the UI thread via Web Socket
- In various cases there are a LOT of messages coming in. 
  - `resource.created`, `resource.change`, `resource.remove` events for resources that change
  - resources that are created/remove cause resource.change to `counts` type (note BE have fixed an issue where `counts` updated even if there are no changes)
- Each message, or 1 seconds worth of batched messages, needs to be processed and stored. Once stored watchers and possible screen renders are triggered.
- This all happens in the UI thread which can struggle

**Describe the solution you'd like**
- Move the handling of the cluster store's socket into a Web Worker
- Optimise how these are sent to the UI thread
  - Batch similar messages
  - Discard changes that are superseded by newer events

**Additional context**
- This is change number 1 of 4
- Step 2 is https://github.com/rancher/dashboard/issues/7895
- Step 3 is https://github.com/rancher/dashboard/issues/7896
- Step 4 is https://github.com/rancher/dashboard/issues/6541

### PR DESCRIPTION
**Pull request by Sean-McQ**
### Summary
Fixes #7894

### CODE REVIEW

File: shell/utils/resourceWatcher.js
Comment (richard-cox) [inline shell/utils/resourceWatcher.js:1]: This needs some comments/docs about what this does and why

File: shell/utils/resourceWatcher.js
Comment (richard-cox) [inline shell/utils/resourceWatcher.js:200]: Are the errors cleared anywhere?

File: shell/utils/resourceWatcher.js
Comment (richard-cox) [inline shell/utils/resourceWatcher.js:206]: We still need to the resyncWatch functionality here

File: shell/utils/resourceWatcher.js
Comment (richard-cox) [inline shell/utils/resourceWatcher.js:76]: This skips boilerplate stuff with have in $axios, specifically the csrf header. Needs investigating why this doesn't fail

File: shell/utils/resourceWatcher.js
Comment (richard-cox) [inline shell/utils/resourceWatcher.js:83]: Should we continue if we've failed to get the revision, won't resourceVersionTime be tracked but resourceVersionTime supplied?

**Review by richard-cox — CHANGES_REQUESTED** [review]
- `shell/plugins/steve/subscribe.js`
  - some docs/comments at the top is needed to explain that subscribe handles resource socket things directly with sockets, or via worker (basic and advanced)
  - `'ws.resource.change'(ctx, msg)` contains an if block that has been removed in master - #7647 (which removed special handling of COUNTS given changes in the backend). This PR needs aligning with that one
  - `'ws.resource.change'(ctx, msg)` & `'ws.resource.remove'(ctx, msg)` have additional code to handle resource aliases which is skipped if the advanced worker is used
  - It should be a little clearer which fns apply to which mode (socket, basic worker, advanced worker). I think this pretty much means the following t
    - Affect following types - rehydrateSubscribe, reconnectWatches, opened, closed, error, send, sendImmediate, `ws.resource.stop`, `ws.resource.start`, `ws.resource.error`
    - Could do this by doc, or throw exception if mode is advanced worker?
  - To confirm, we don't need to do anything special for resyncWatch?
  - How is `ws.ping` handled in the advanced worker? (update - doesn't look like this is passed through from advanced worker)
- I'm still not sure I understand why `resourceWatcher` needs to update state on an interval instead of directly when needed. it introduces a lot of complexity (and delay) and the benefit isn't clear. 
- I've only given this a brief test but hit two issue. 
  - issue 1 
    - In browser 1 bring up the Deployments list
    - In browser 2 go to same list, edit a deployment and scale it up 
    - In browser 1 the deployment stats should change, however remain with the previous value
  - issue 2 
    - switching between clusters should close and open the socket
    - (probably due to unsubscribe not being wired in)
  
My TODO's
- All advanced worker testing
- Test normal socket behaviour plus basic worker

File: shell/plugins/dashboard-store/mutations.js
Comment (Sean-McQ) [reply shell/plugins/dashboard-store/mutations.js:217]: I did go ahead and move the map add/setting into the reduce function instead of looping over the array again.

If the limit is lower than the count the slice is a simple no-op and shouldn't impact the complexity of the mutation at all

You're right, I was using addObjects just because that's what we were doing in other mutations but it's more than I need here so I swapped it out for a simple push.

I could further refine this, by using the index on the cache to directly modify indexes, add new resources, and then slice out the removed ones but I don't think we'd be getting much by cutting out a linear loop over the array (at the end of the day, simple loops over an array that's only in the 10s of thousands is in the 10s of milliseconds) and done is better than perfect here probably.

File: shell/plugins/steve/subscribe.js
Comment (Sean-McQ) [reply shell/plugins/steve/subscribe.js:362]: it'll always be false and therefore the watch will always eventually make it to the worker where it'll check to see if it's watching or not there. That being said, I went ahead and made some modifications so that when we start a watch we signal back to the UI thread so the watchStarted getter will be accurate.

File: shell/plugins/dashboard-store/mutations.js
Comment (Sean-McQ) [reply shell/plugins/dashboard-store/mutations.js]: this is now an exported function from the schema model that we can use elsewhere as well. I'll need it in the resourceCaches later as well.

File: shell/utils/resourceWatcher.js
Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: moved it into `@shell/plugins/steve' to specify that it's a steve-specific resourceWatcher.

File: shell/plugins/dashboard-store/mutations.js
Comment (Sean-McQ) [reply shell/plugins/dashboard-store/mutations.js]: done

File: shell/plugins/dashboard-store/mutations.js
Comment (Sean-McQ) [reply shell/plugins/dashboard-store/mutations.js]: I don't see any other mutations actually using the getter classify here or for that matter in any mutations so I'd either need to modify all mutations that use classify or leave this one as is to maintain consistency.

Comment (Sean-McQ) [reply shell/plugins/dashboard-store/mutations.js]: that makes sense, I was classifying again just to make sure I also classified the added resources as well, seemed relatively cheap since the function actually returns immediately if the object's type is already a resource but it's just as easy to classify the created resources when I push them into their array.

File: shell/plugins/steve/subscribe.js
Comment (Sean-McQ) [reply shell/plugins/steve/subscribe.js]: done, there's now an async utils file.

File: shell/plugins/steve/worker/index.js
Comment (Sean-McQ) [reply shell/plugins/steve/worker/index.js]: done, my local linting won't shut up about this if I don't have those there :)

File: shell/plugins/steve/subscribe.js
Comment (Sean-McQ) [reply shell/plugins/steve/subscribe.js]: it didn't, I added some logic for the advanced worker to pass certain events back to the UI thread and among other things, it should properly unsubscribe now.

File: shell/plugins/steve/subscribe.js
Comment (Sean-McQ) [reply shell/plugins/steve/subscribe.js]: done although I tend to like it better at the top as I think declaring constants early makes the code more readable but it works just the same.

File: shell/plugins/steve/worker/web-worker.advanced.js
Comment (Sean-McQ) [reply shell/plugins/steve/worker/web-worker.advanced.js]: yup, I had dialed it down to 1 second for testing and forgot to dial it back.

File: shell/plugins/steve/worker/web-worker.advanced.js
Comment (Sean-McQ) [reply shell/plugins/steve/worker/web-worker.advanced.js]: done, I removed the resourceWatcherAction which means that the worker will now dispatch it back to the UI thread where it'll be dispatched as though it came in normally as opposed to the worker.

File: shell/plugins/steve/subscribe.js
Comment (Sean-McQ) [reply shell/plugins/steve/subscribe.js]: good catch, fixed.

File: shell/utils/resourceWatcher.js
Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: added some comments, full docs should probably wait until the whole thing is implemented to avoid rewriting as much as possible.

File: shell/utils/resourceWatcher.js
Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: it'll be cleared upon a successful watch when I rewrite the watchObject

File: shell/utils/resourceWatcher.js
Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: unless I'm missing something, I think the watch function here should cover everything that needs to happen on "too old" for us, including getting the new revision.

File: shell/utils/resourceWatcher.js
Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: I think I added this at your recommendation (I could be thinking of something different though), I went ahead and removed it.

File: shell/utils/resourceWatcher.js
Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: it doesn't fail because fetch automatically inserts a cookie header which I verified by logging calls outside of the dev-tools (for whatever reason, devtools doesn't like showing the headers properly here). If need be, I can inject the CSRF cookie value into the worker from the UI thread but I can't directly get to the cookie from inside of the worker which means I'll always have to push it in. That being said, the value only changes when the session changes which would typically mean the page is at least refreshed which means I'm rebuilding the worker anyways so keeping it synced shouldn't be too big a deal.

File: shell/utils/resourceWatcher.js
Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: there's some resources that simply don't return a revision (nodemetrics being one of them I think) so I don't want the socket to just die on the vine here if that's the case, especially if there wasn't an error. I added a flag to the watch to keep track of that case though so we can use the error to know how the fetch failed.

File: shell/plugins/dashboard-store/mutations.js
Comment (richard-cox) [reply shell/plugins/dashboard-store/mutations.js]: I had this backwards, you're doing it correctly. The flow is 

classify fn --> classify getter --> lookup fn

where i thought it was

classify getter --> classify fn --> --> lookup fn

File: shell/utils/resourceWatcher.js
Comment (richard-cox) [reply shell/utils/resourceWatcher.js]: `watch` will only fetch a new revision if the resourveVersion states are cleared or invalid. We're going on the assumption it's 5 minutes, depending on server setting could skip if it's not hit.

also `resyncWatch` removes stale entries

Comment (Sean-McQ) [reply shell/utils/resourceWatcher.js]: so this'll now clear out the resourceVersion and resourceVersionTime which should force it to go get a new one.

I also set this to emit a resyncWatch event message after the watch is completed which will be passed from the worker to the UI thread.

**Comment by richard-cox** [comment]
# Summary of Dev Testing

## To Do
### Settings/Old Way
- Disable the feature, ensure similar all other tests (or subset) work fine

### Specific types of events
- Ensure different type of watches work fine (resource.start, resource.stop, resource.change events all work as expected)
  - selector
  - namespace
  - id
### Disconnect/Reconnect
- Disconnect socket / socket down --> nav to a new resouce (aka watch is queued). Socket comes up. Confirm `resource.start` message successfully received for new resource

### Regression
- #5997 - Ensure that when steve resource stops watching we successfully re-watch with appropriate revision


## Complete
### Resources update
- Browser 1 deployment detail page, pods list visible. Browser 2 --> delete a pod. Browser 1 pods list should show change new pod, remove old pod including any change of pod states
  - Same as above, but redeploy deployment from deployment list
### Batch updates work as expected
- Cluster dashboard, events list visible, total events is at 500 and no more
  - Note - if there's not enough events to hit the events limit, edit `configureType(EVENT, { limit: 500 });)`
### Specific types of events
- Cluster dashboard --> any other cluster page. Certain types should be forgotten aka unwatch called resulting in `resource.stop` socket message
  - Returning to the cluster dashboard should re-watch same types aka `resource.start` socket message 
- Cluster dashboard, events list visible. Click on an event and confirm `events` `resource.stop` is received and `events` for specific event id is `resource.start`ed. Return to cluster dashboard and ensure reverse happens
### Disconnect/Reconnect
- Browser 1 cluster dashboard. Browser 2 on local cluster, redeploy the `cattle-cluster-agent`. cluster socket should close, socket re-connect messages in console should appear in reducing amount. when agent comes back up cluster socket should open and stay open. Note - there should be messages to re-sub to anything that was sub'd... however all will fail. This is the same as on master
  - Same as above, but kill the pod within the `cattle-cluster-agent` deployment

**Review by richard-cox — APPROVED** [review]
Given the below is done this is good to ***squash merge***
- Remaining tests from https://github.com/rancher/dashboard/pull/7760#issuecomment-1380211326
  - One is blocked by https://github.com/rancher/dashboard/issues/7919
- @Sean-McQ Can you give the code a brief self review before merging, just to catch any bad logging, settings, etc?

> Work for next release in https://github.com/rancher/dashboard/issues/7917

**Comment by catz** [comment]
Hello @richard-cox,

I've been checking the code in the advanced worker and noticed that there seems to be a missing piece for the COUNT entity (https://github.com/rancher/dashboard/blob/479074f5560a3f7a0813909f56e5fe17567be53d/shell/plugins/steve/worker/web-worker.advanced.js#L238).

I suppose this works fine for standard resources, but not for counts. We need to merge previously saved counts; otherwise, only the last counts will be included in the batch changes to send.

(Perhaps something similar to this could work: https://github.com/rancher/dashboard/pull/7647)

**Pull request merged**
